model/riscv_types_crypto.sail

```diff
@@ -0,0 +1,522 @@
+
+/*
+ * file: riscv_crypto_types.sail
+ *
+ * This file contains types, mappings and functions used across the
+ * cryptography extension instructions.
+ *
+ * This file must be included in the model build whatever the value of XLEN.
+ *
+ */
+
+
+/*
+ * Cryptography extension types.
+ * ----------------------------------------------------------------------
+ */
+
+/* Byte shift parameter for scalar 32-bit aes instructions */
+type            aes_bs = bits(2)
+
+/* Scalar AES round instruction operations. Used for 32,64-bit versions */
+enum            aes_op = {
+    ENCS, ENCSM, DECS, DECSM
+}
+
+
+/* Scalar SHA256 instruction operations. Used for 32,64-bit versions. */
+enum            sha256_op = {
+    OP_SHA256_SIG0,
+    OP_SHA256_SIG1,
+    OP_SHA256_SUM0,
+    OP_SHA256_SUM1
+}
+
+/* Scalar SHA512 RV64 instruction operations */
+enum            sha512_rv64_op = {
+    OP_SHA512_SIG0,
+    OP_SHA512_SIG1,
+    OP_SHA512_SUM0,
+    OP_SHA512_SUM1
+}
+
+/* Scalar SHA512 RV32 instruction operations.*/
+enum            sha512_rv32_op = {
+    OP_SHA512_SIG0L, OP_SHA512_SIG0H,
+    OP_SHA512_SIG1L, OP_SHA512_SIG1H,
+    OP_SHA512_SUM0R, OP_SHA512_SUM1R
+}
+
+/* Scalar SM3 instruction operations. Used for 32,64-bit versions. */
+enum            sm3_op = {
+    P0, P1
+}
+
+/* Scalar SM4 instruction operations. Used for 32,64-bit versions. */
+enum            sm4_op = {
+    KS, ED
+}
+
+/* Map scalar instruction round function ops to whether they enc or dec? */
+mapping aes_op_fwd        : aes_op <-> bool = {
+    ENCSM   <-> true    ,
+    ENCS    <-> true    ,
+    DECSM   <-> false   ,
+    DECS    <-> false
+}
+
+/* Map scalar instruction round function ops to whether they perform mix? */
+mapping aes_op_does_mix   : aes_op <-> bool = {
+    ENCSM   <-> true    ,
+    ENCS    <-> false   ,
+    DECSM   <-> true    ,
+    DECS    <-> false 
+}
+
+
+/* 
+ * Cryptography extension shared / utility functions
+ * ----------------------------------------------------------------------
+ */
+
+val      ror32  : (bits(32), int) -> bits(32)
+function ror32  (x, y) = {
+    (x>>to_bits(5,y)) | (x<<to_bits(5,32-y))
```

Comment:
Whitespace

Reply:
Please forgive my asking - you want all extra whitespace removed? E.g. between the commas on lines 70-73 and the previous token?

Reply:
No, there isn't enough (binary operators want spaces around them, commas want spaces after them).

Reply:
Great, glad I asked!